services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java

```diff
@@ -6465,6 +6465,94 @@ public boolean isLeaderControllerFor(String clusterName) {
     return model.isLeader();
   }
 
+  @Override
+  public InstanceRemovableStatuses getAggregatedHealthStatus(
+      String cluster,
+      Set<String> instances,
+      List<String> toBeStoppedInstances) {
+    InstanceRemovableStatuses statuses = new InstanceRemovableStatuses();
+
+    // If current controller is not the leader, redirect with leader URL
+    if (!isLeaderControllerFor(cluster)) {
+      Instance instance = getLeaderController(cluster);
+      statuses.setRedirectUrl(instance.getNodeId());
+      return statuses;
+    }
+
+    Map<String, String> nonStoppableInstances = new HashMap<>();
+    List<String> stoppableInstances = new ArrayList<>();
+    statuses.setNonStoppableInstances(nonStoppableInstances);
+    statuses.setStoppableInstances(stoppableInstances);
+
+    // Check for maintenance, for ongoing maintenance none can be removed.
+    HelixVeniceClusterResources resources = getHelixVeniceClusterResources(cluster);
+    MaintenanceSignal maintenanceSignal = HelixUtils.getClusterMaintenanceSignal(cluster, resources.getHelixManager());
+
+    if (maintenanceSignal != null) {
+      for (String instanceId: instances) {
+        nonStoppableInstances.put(instanceId, InstanceRemovableStatuses.NonStoppableReason.ONGOING_MAINTENANCE.name());
+      }
+      return statuses;
+    }
+
+    List<String> allInstances = helixAdminClient.getInstancesInCluster(cluster);
+    Set<Instance> liveInstances = getLiveInstanceMonitor(cluster).getAllLiveInstances();
+
+    HelixCustomizedViewOfflinePushRepository customizedViewRepo = resources.getCustomizedViewRepository();
+    ReadWriteStoreRepository storeRepo = getHelixVeniceClusterResources(cluster).getStoreMetadataRepository();
+
+    for (String instanceId: instances) {
```

Comment:
Can we unify the logic b/w this and `node_removable` checks? That one already has the functionality of `to_be_stopped_instances` through `LOCKED_STORAGE_NODE_IDS`

Reply:
May be not in this PR.  `node_removable` does not use CV and calling it in aggregate over a batch of instances require internal logic change.

Reply:
We can still abstract the main code in one method, and call it from two places. `node_removable` can pass a single-element-list to this new function.

Also, this method can accept a `RoutingDataRepository`, and `stoppable-check` can pass CV repo while `node_removable` can pass EV repo